File: android/src/main/java/com/nabbra/rnbluesmanager/models/BluetoothNativeDevice.kt
package com.nabbra.rnbluesmanager.models

import android.annotation.SuppressLint
import android.bluetooth.BluetoothClass
import android.bluetooth.BluetoothDevice
import android.os.ParcelUuid
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.WritableMap
import com.nabbra.rnbluesmanager.contracts.MappableContract

@SuppressLint("MissingPermission")
class BluetoothNativeDevice(private val device: BluetoothDevice): MappableContract {
  private val extra: MutableMap<String, Any> = mutableMapOf()

  val address: String
    get() = device.address

  val name: String?
    get() = device.name

  val bondState: Int
    get() = device.bondState

  val bluetoothClass: BluetoothClass?
    get() = device.bluetoothClass

  val uuids: Array<ParcelUuid>?
    get() = device.uuids

  fun getDevice(): BluetoothDevice = device

  fun getExtra(key: String): Any? = extra[key]

  fun putExtra(key: String, value: Any): Any? = extra.put(key, value)

  override fun map(): WritableMap? {
    val mapped = Arguments.createMap().apply {
      putString("name", device.name ?: device.address)
      putString("address", device.address)
      putString("id", device.address)
      putBoolean("bonded", device.bondState == BluetoothDevice.BOND_BONDED)

      device.bluetoothClass?.let { btClass ->
        putMap("deviceClass", Arguments.createMap().apply {
          putInt("deviceClass", btClass.deviceClass)
          putInt("majorClass", btClass.majorDeviceClass)
        })
      }

      putMap("extra", Arguments.createMap().apply {
        extra.forEach { (key, value) ->
          when (value) {
            is Int -> putInt(key, value)
            is String -> putString(key, value)
            is Boolean -> putBoolean(key, value)
            is Double -> putDouble(key, value)
            else -> putString(key, value.toString())
          }
        }
      })
    }

    return mapped
  }

  override fun toString(): String = map().toString()
}

File: ios/models/BluetoothNativeDevice.swift
import Foundation
import AVFoundation
import React

@objcMembers
class BluetoothNativeDevice: NSObject {
    private var nameValue: String
    private var addressValue: String
    private var bondedValue: Bool
    private var extra: [String: Any] = [:]
    
    init(name: String, address: String, bonded: Bool = false) {
        self.nameValue = name
        self.addressValue = address
        self.bondedValue = bonded
        super.init()
    }
    
    var name: String { nameValue }
    var address: String { addressValue }
    var id: String { addressValue }
    var bonded: Bool { bondedValue }
    
    func getExtra(key: String) -> Any? {
        return extra[key]
    }

    func putExtra(key: String, value: Any) -> Any? {
        return extra.updateValue(value, forKey: key)
    }

    func map() -> [String: Any] {
        var mapped: [String: Any] = [
            "name": nameValue,
            "address": addressValue,
            "id": id,
            "bonded": bondedValue
        ]
        mapped["extra"] = extra
        return mapped
    }

    override var description: String {
        return "\(map())"
    }
}
